# Explore here

## Cell 2: Load dataset
This cell imports pandas and loads the Bank Marketing dataset from a public CSV URL.

What this cell does:
- Reads the file using semicolon separator.
- Stores the full dataset in total_data.
- Displays the first five rows to verify schema and values.

Expected output:
A preview table with the first records of the raw dataset.

In [60]:
# Your code here
import pandas as pd

total_data = pd.read_csv("https://raw.githubusercontent.com/4GeeksAcademy/logistic-regression-project-tutorial/main/bank-marketing-campaign-data.csv", sep = ";")
total_data.head()

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
0,56,housemaid,married,basic.4y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
1,57,services,married,high.school,unknown,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
2,37,services,married,high.school,no,yes,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
3,40,admin.,married,basic.6y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
4,56,services,married,high.school,no,no,yes,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no


## Cell 3: Remove duplicate rows
This cell performs basic data cleaning by removing exact duplicate records.

What this cell does:
- Applies drop_duplicates to keep one copy of repeated rows.
- Resets row indexing so the DataFrame remains sequential.
- Shows the first five rows of the cleaned dataset.

Why this matters:
Duplicate records can bias model training and evaluation.

In [61]:
total_data = total_data.drop_duplicates().reset_index(drop = True)
total_data.head()

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
0,56,housemaid,married,basic.4y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
1,57,services,married,high.school,unknown,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
2,37,services,married,high.school,no,yes,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
3,40,admin.,married,basic.6y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
4,56,services,married,high.school,no,no,yes,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no


## Cell 4: Check missing values
This cell audits data completeness before feature engineering.

What this cell does:
- Computes missing-value counts for every column.
- Returns a per-column summary using isnull().sum().

How to interpret:
- Zero values mean no missing entries in that column.
- Non-zero values indicate columns that may need imputation or dropping.

In [62]:
total_data.isnull().sum()

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
y                 0
dtype: int64

## Cell 5: Encode and scale features
This cell converts categorical columns into numeric `_n` columns and scales selected variables.

What this cell does:
- Uses `pd.factorize()` to encode categorical fields into integer codes (`job_n`, `marital_n`, `education_n`, ..., `y_n`).
- Builds a `num_variables` list with encoded categorical features and original numeric features.
- Applies `MinMaxScaler` to normalize all variables in `num_variables` to the range `[0, 1]`.
- Creates `total_data_scal` as the transformed dataset.

Final output:
- `total_data_scal` contains the scaled model-ready variables.
- `total_data_scal.head()` displays a preview of transformed data.

In [63]:
from sklearn.preprocessing import MinMaxScaler

total_data["job_n"] = pd.factorize(total_data["job"])[0]
total_data["marital_n"] = pd.factorize(total_data["marital"])[0]
total_data["education_n"] = pd.factorize(total_data["education"])[0]
total_data["default_n"] = pd.factorize(total_data["default"])[0]
total_data["housing_n"] = pd.factorize(total_data["housing"])[0]
total_data["loan_n"] = pd.factorize(total_data["loan"])[0]
total_data["contact_n"] = pd.factorize(total_data["contact"])[0]
total_data["month_n"] = pd.factorize(total_data["month"])[0]
total_data["day_of_week_n"] = pd.factorize(total_data["day_of_week"])[0]
total_data["poutcome_n"] = pd.factorize(total_data["poutcome"])[0]
total_data["y_n"] = pd.factorize(total_data["y"])[0]
num_variables = ["job_n", "marital_n", "education_n", "default_n", "housing_n", "loan_n", "contact_n", "month_n", "day_of_week_n", "poutcome_n",
                 "age", "duration", "campaign", "pdays", "previous", "emp.var.rate", "cons.price.idx", "cons.conf.idx", "euribor3m", "nr.employed", "y_n"]

scaler = MinMaxScaler()
scal_features = scaler.fit_transform(total_data[num_variables])
total_data_scal = pd.DataFrame(scal_features, index = total_data.index, columns = num_variables)
total_data_scal.head()

,job_n,marital_n,education_n,default_n,housing_n,loan_n,contact_n,month_n,day_of_week_n,poutcome_n,...,duration,campaign,pdays,previous,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y_n
0,0.000000,0.0,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.053070,0.0,1.0,0.0,0.9375,0.698753,0.60251,0.957379,0.859735,0.0
1,0.090909,0.0,0.142857,0.5,0.0,0.0,0.0,0.0,0.0,0.0,...,0.030297,0.0,1.0,0.0,0.9375,0.698753,0.60251,0.957379,0.859735,0.0
2,0.090909,0.0,0.142857,0.0,0.5,0.0,0.0,0.0,0.0,0.0,...,0.045954,0.0,1.0,0.0,0.9375,0.698753,0.60251,0.957379,0.859735,0.0
3,0.181818,0.0,0.285714,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.030704,0.0,1.0,0.0,0.9375,0.698753,0.60251,0.957379,0.859735,0.0
4,0.090909,0.0,0.142857,0.0,0.0,0.5,0.0,0.0,0.0,0.0,...,0.062424,0.0,1.0,0.0,0.9375,0.698753,0.60251,0.957379,0.859735,0.0


## Cell 6: Feature selection and train/test split
This cell prepares modeling datasets using chi-square-based feature selection.

What this cell does:
- Splits scaled data into features (`X`) and target (`y_n`).
- Creates train/test sets with `test_size = 0.1` and `random_state = 42`.
- Uses `SelectKBest(chi2, k = 5)` fitted on training data only.
- Applies the same selector to both train and test sets.
- Creates `X_train_sel` and `X_test_sel` with selected feature names.

Final output:
- `X_train`, `X_test`, `y_train`, `y_test` are available.
- `X_train_sel` and `X_test_sel` contain the top 5 selected features.
- `X_train_sel.head()` previews the selected training features.

In [64]:
# Feature selection (Chi-square) + train/test split

from sklearn.feature_selection import chi2, SelectKBest
from sklearn.model_selection import train_test_split


X = total_data_scal.drop("y_n", axis = 1)
y = total_data_scal["y_n"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.1, random_state = 42)

selection_model = SelectKBest(chi2, k = 5)
selection_model.fit(X_train, y_train)
ix = selection_model.get_support()
X_train_sel = pd.DataFrame(selection_model.transform(X_train), columns = X_train.columns.values[ix])
X_test_sel = pd.DataFrame(selection_model.transform(X_test), columns = X_test.columns.values[ix])

X_train_sel.head()

,poutcome_n,previous,emp.var.rate,euribor3m,nr.employed
0,0.0,0.0,1.000000,0.980730,1.000000
1,0.0,0.0,0.333333,0.138291,0.512287
2,0.0,0.0,0.937500,0.956926,0.859735
3,0.0,0.0,0.937500,0.957379,0.859735
4,0.0,0.0,0.333333,0.175924,0.512287


## Cell 7: Preview selected test features
This cell displays a preview of the selected features in the test split.

Expected output:
A small table showing `X_test_sel.head()`.

In [65]:
X_test_sel.head()

,poutcome_n,previous,emp.var.rate,euribor3m,nr.employed
0,0.0,0.000000,0.104167,0.143278,0.425709
1,0.5,0.142857,0.333333,0.150759,0.512287
2,0.0,0.000000,0.479167,0.073679,0.000000
3,0.5,0.142857,0.333333,0.150759,0.512287
4,0.0,0.000000,1.000000,0.981410,1.000000


## Cell 8: Save processed train and test data
This cell appends target labels to selected features and exports clean CSV files for modeling.

Expected output:
`clean_train.csv` and `clean_test.csv` saved in `../data/processed/`.

In [66]:
X_train_sel["y_n"] = list(y_train)
X_test_sel["y_n"] = list(y_test)
X_train_sel.to_csv("../data/processed/clean_train.csv", index = False)
X_test_sel.to_csv("../data/processed/clean_test.csv", index = False)

## Cell 9: Load processed datasets
This cell loads the saved processed train and test files into DataFrames for model training.

Expected output:
Preview of `train_data.head()`.

In [67]:
train_data = pd.read_csv("../data/processed/clean_train.csv")
test_data = pd.read_csv("../data/processed/clean_test.csv")

train_data.head()

,poutcome_n,previous,emp.var.rate,euribor3m,nr.employed,y_n
0,0.0,0.0,1.000000,0.980730,1.000000,0.0
1,0.0,0.0,0.333333,0.138291,0.512287,0.0
2,0.0,0.0,0.937500,0.956926,0.859735,0.0
3,0.0,0.0,0.937500,0.957379,0.859735,0.0
4,0.0,0.0,0.333333,0.175924,0.512287,0.0


## Cell 10: Train baseline logistic regression
This cell trains a baseline logistic regression model and predicts labels on the test set.

Expected output:
`y_pred` array with predicted class labels.

In [68]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, roc_auc_score

# Split features and target
X_train_model = train_data.drop(columns=["y_n"])
y_train_model = train_data["y_n"].astype(int)

X_test_model = test_data.drop(columns=["y_n"])
y_test_model = test_data["y_n"].astype(int)

# Train model
logistic_model = LogisticRegression(max_iter=1000)
logistic_model.fit(X_train_model, y_train_model)

# Predict
y_pred = logistic_model.predict(X_test_model)
y_pred


array([0, 0, 0, ..., 0, 0, 0], shape=(4118,))

## Cell 11: Evaluate baseline accuracy
This cell calculates the baseline test accuracy for the logistic regression model.

Expected output:
A single accuracy score between 0 and 1.

In [69]:
from sklearn.metrics import accuracy_score

accuracy_score(y_test_model, y_pred)

0.8919378338999514

## Cell 12: Hyperparameter tuning with GridSearchCV
This cell runs cross-validated grid search to find the best logistic regression hyperparameters.

Expected output:
Best parameter combination and best cross-validation accuracy.

In [70]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression

param_grid = [
    {
        "solver": ["newton-cg", "lbfgs", "sag"],
        "penalty": ["l2"],
        "C": [0.001, 0.01, 0.1, 1, 10, 100, 1000]
    },
    {
        "solver": ["liblinear"],
        "penalty": ["l1", "l2"],
        "C": [0.001, 0.01, 0.1, 1, 10, 100, 1000]
    },
    {
        "solver": ["saga"],
        "penalty": ["l1", "l2", "elasticnet"],
        "C": [0.001, 0.01, 0.1, 1, 10, 100, 1000],
        "l1_ratio": [0.0, 0.5, 1.0]
    }
]

model = LogisticRegression(max_iter = 2000, random_state = 42)
grid = GridSearchCV(model, param_grid, scoring = "accuracy", cv = 5, n_jobs = 1)
grid.fit(X_train_model, y_train_model)
best_model = grid.best_estimator_

print("Best params:", grid.best_params_)
print("Best CV accuracy:", grid.best_score_)

Best params: {'C': 0.1, 'penalty': 'l2', 'solver': 'lbfgs'}
Best CV accuracy: 0.8971072266593982


## Cell 13: Predict with tuned model
This cell uses the tuned model to predict labels on the test set.

Expected output:
`y_pred_tuned` array with predicted labels.

In [71]:
y_pred_tuned = best_model.predict(X_test_model)
y_pred_tuned

array([0, 0, 0, ..., 0, 0, 0], shape=(4118,))

## Cell 14: Evaluate tuned model
This cell computes test accuracy for tuned model predictions.

Expected output:
Tuned model accuracy score.

In [72]:
accuracy_score(y_test_model, y_pred_tuned)

0.8938805245264692

## Cell 15: Save trained model
This cell serializes the tuned model to the `models` directory for reuse.

Expected output:
A `.sav` model file written to `../models/`.

In [73]:
from pickle import dump

dump(best_model, open("../models/logistic_regression_tuned.sav", "wb"))

## Final Summary: Baseline vs Tuned Logistic Regression

### Results comparison
- Baseline test accuracy: **0.8919**
- Tuned model cross-validation accuracy (best CV): **0.8971**
- Tuned model test accuracy: **0.8939**

### Interpretation
- Hyperparameter tuning produced a **small improvement** over the baseline on the test set.
- The tuned model is saved as `../models/logistic_regression_tuned.sav` and can be reused for inference.